core/arch/arm/mm/mobj.c

```diff
@@ -639,6 +663,68 @@ TEE_Result mobj_reg_shm_unmap(struct mobj *mobj)
 	return TEE_SUCCESS;
 }
 
+TEE_Result mobj_reg_shm_inc_map(struct mobj *mobj)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct mobj_reg_shm *r = to_mobj_reg_shm_may_fail(mobj);
+
+	if (!r)
+		return TEE_ERROR_GENERIC;
+
+	if (refcount_inc(&r->mapcount))
+		return TEE_SUCCESS;
+
+	uint32_t exceptions = cpu_spin_lock_xsave(&reg_shm_map_lock);
```

Comment:
Thread A gets a mapcount, does something then releases the mapping: it calls `mobj_reg_shm_dec_map()` below which decrements the mapcount but, at this time, the CPU gets an IRQ before thread has the spinlock at (line ~713).
`map counter = 0  &&  map state = area still mapped`

Thread B arrives, increments map count from 0, maps the area and do its duty.
`  map counter = 1 && map state = area mapped`

Thread A comes back on line, gets the spinlock (line ~713) and completes unmap sequence.
` =>` area is unmaped

Do I miss something?

Reply:
If counter reaches 0 it will not be increased again, it will have to be reinitialized. See https://github.com/OP-TEE/optee_os/blob/52a5324c62d0a19875278b9ceff945c74f95b609/core/kernel/refcount.c#L18-L20
also see
https://github.com/OP-TEE/optee_os/blob/52a5324c62d0a19875278b9ceff945c74f95b609/core/include/kernel/refcount.h#L11-L48

By checking the value of the counter after taking the spinlock we make sure we have the correct view of the state before modifying it.

Reply:
ok. thanks the details. crystal clear.